File: backend/app/schemas/projects.py
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


ProjectStatus = Literal["ACTIVE", "PAUSED", "COMPLETED", "ARCHIVED"]
ProjectTitle = Annotated[str, Field(min_length=1, max_length=160)]
ProjectObjective = Annotated[str, Field(min_length=1, max_length=4000)]
ChangeNote = Annotated[str, Field(min_length=1, max_length=512)]


class ChangeNotedRequest(BaseModel):
    change_note: ChangeNote

    @field_validator("change_note")
    @classmethod
    def normalize_change_note(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("change_note must contain non-whitespace text.")
        return normalized


class CreateProjectRequest(ChangeNotedRequest):
    title: ProjectTitle
    objective: ProjectObjective
    change_note: ChangeNote = "Owner created project."


class UpdateProjectDetailsRequest(ChangeNotedRequest):
    expected_revision: int = Field(ge=1)
    title: ProjectTitle | None = None
    objective: ProjectObjective | None = None


class RecordProjectProgressRequest(ChangeNotedRequest):
    expected_revision: int = Field(ge=1)
    current_summary: str | None = Field(default=None, max_length=4000)


class ChangeNextActionRequest(ChangeNotedRequest):
    expected_revision: int = Field(ge=1)
    next_action: str | None = Field(default=None, max_length=512)
    @field_validator("next_action")
    @classmethod
    def reject_whitespace_next_action(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("next_action must be non-whitespace text or null.")
        return value.strip() if value is not None else None


class ChangeProjectStatusRequest(ChangeNotedRequest):
    expected_revision: int = Field(ge=1)
    status: ProjectStatus


class ProjectRevisionResponse(BaseModel):
    id: UUID
    revision_number: int
    title: str
    objective: str
    status: ProjectStatus
    current_summary: str | None
    next_action: str | None
    change_note: str
    created_at: datetime


class ProjectResponse(BaseModel):
    id: UUID
    title: str
    objective: str
    status: ProjectStatus
    current_summary: str | None
    next_action: str | None
    current_revision: int
    created_at: datetime
    updated_at: datetime


class ProjectListResponse(BaseModel):
    items: list[ProjectResponse]
    page: int
    page_size: int
    total: int


class ProjectHistoryResponse(BaseModel):
    items: list[ProjectRevisionResponse]

File: frontend/types/projects.ts
export type ProjectStatus = "ACTIVE" | "PAUSED" | "COMPLETED" | "ARCHIVED";

export interface Project {
  id: string;
  title: string;
  objective: string;
  status: ProjectStatus;
  current_summary: string | null;
  next_action: string | null;
  current_revision: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectRevision {
  id: string;
  revision_number: number;
  title: string;
  objective: string;
  status: ProjectStatus;
  current_summary: string | null;
  next_action: string | null;
  change_note: string;
  created_at: string;
}

export interface ProjectListResponse {
  items: Project[];
  page: number;
  page_size: number;
  total: number;
}

export interface ProjectHistoryResponse {
  items: ProjectRevision[];
}

export interface CreateProjectRequest {
  title: string;
  objective: string;
  change_note: string;
}

export interface UpdateProjectDetailsRequest {
  expected_revision: number;
  change_note: string;
  title?: string;
  objective?: string;
}

export interface RecordProjectProgressRequest {
  expected_revision: number;
  change_note: string;
  current_summary: string | null;
}

export interface ChangeNextActionRequest {
  expected_revision: number;
  change_note: string;
  next_action: string | null;
}

export interface ChangeProjectStatusRequest {
  expected_revision: number;
  change_note: string;
  status: ProjectStatus;
}
